Add a batch_first arg to Transformer / MHA modules

scaomath opened the pull request:
Fixes #25100 #43112

EDIT: pardon my inexperience since this is my first PR here, that I did not realize the doc should not have any trailing white spaces, and `[E712] comparison to False should be 'if cond is False:' or 'if not cond:'`, now both fixed.

scaomath committed:
update batch_first for transformers, encoder, decoder layers, MHA
update test_nn to add batch_first block, adjust test_nn atol for inputs with different magnitudes

zhangguanheng66 commented on torch/nn/modules/transformer.py:
```diff
@@ -227,7 +246,11 @@ def forward(self, tgt: Tensor, memory: Tensor, tgt_mask: Optional[Tensor] = None
             see the docs in Transformer class.
         """
-        output = tgt
+        if self.batch_first:
```
Do we want to have the transpose on the transformer level or the MHA level?

scaomath replied:
Locally on my end, I followed implementation here [https://nlp.seas.harvard.edu/2018/04/03/attention.html](https://nlp.seas.harvard.edu/2018/04/03/attention.html) and changed the batch first in the MHA level so that the input to MHA is `(batch, n_head, seq_len, d_model//n_head)`. But I thought it would be simpler if just change it in the wrapper level.

jbschlosser replied:
I think it's fairly common to use MHA directly without the Transformer layers, since they're not flexible enough yet for certain use cases. Given that, I think it'd be good to support `batch_first` at the MHA level.

jbschlosser reviewed (commented):
Thanks for your contribution! As mentioned, would you mind updating MHA and TransformerEncoderLayer / TransformerDecoderLayer as well? Additionally, please add some tests into `test/test_nn.py` to validate the behavior for the new flag.

scaomath commented:
> Thanks for your contribution! As mentioned, would you mind updating MHA and TransformerEncoderLayer / TransformerDecoderLayer as well? Additionally, please add some tests into `test/test_nn.py` to validate the behavior for the new flag.

Alright, `batch_first` added to MHA and the layer wrappers too. Some tests added as well. 

> Thanks for your contribution! As mentioned, would you mind updating MHA and TransformerEncoderLayer / TransformerDecoderLayer as well? Additionally, please add some tests into `test/test_nn.py` to validate the behavior for the new flag.

Alright `batch_first` is added to the layer and MHA wrappers as well, I did not touch the base in `functional`.

jbschlosser commented on torch/nn/modules/activation.py:
```diff
@@ -868,7 +870,7 @@ class MultiheadAttention(Module):
     bias_k: Optional[torch.Tensor]
```
Just above this line, please add this line for TorchScript support:

```python
__constants__ = ['batch_first']
```

This should be added to each of the affected modules.

scaomath replied:
Added in `transformer.py` and `activation.py` in affected modules.

jbschlosser reviewed (commented):
Thanks for adding this! Added a couple minor comments.

scaomath commented on test/test_nn.py:
```diff
@@ -5783,6 +5783,39 @@ def test_Transformer_cell(self):
                              memory_key_padding_mask=memory_key_padding_mask)
         output.sum().backward()
 
+    def test_Transformer_cell_batch_first(self):
```
Cool. Now `for batch_first in (True, False):` blocks are used.

scaomath replied:
Now the three errors in the `test_nn.py` are fixed, as I have trouble running directly the `test_nn.py` on my end due to VScode complained submodules not being loaded. I created a minimal environ to make sure the errors are fixed: https://github.com/scaomath/torch-playground

scaomath replied:
I felt that this error is on the verge to be a non-rounding error. On my end, using the original `test_nn.py`, then test cannot pass rel tol 1e-7 and abs tol 1e-5, this is the maximum diff for each entries of `result` and `ref_output` in each test of `test_transformerdecoder`:
```
0.00012551248
1.7166138e-05
8.940697e-07
9.536743e-07
9.536743e-07
9.536743e-07
4.7683716e-07
0.00015902519
2.861023e-06
9.2715025e-05
4.053116e-06
0.00016299635
2.2487715e-05
4.7683716e-07
3.091991e-07
```

Since `test_transformerdecoderlayer` has no problem on `batch_first in (False, True)` loop, I feel the `test_transformerdecoder` cannot pass the test due to other changes.

scaomath replied:
Okay, culprit found when I am preparing my class this morning. It is not reasonable to set the same absolute tolerance for inputs with different magnitudes. For example, the input matrix has entries all between 0 and 1, vs another input that ranges from 40 to 80, of course we are gonna get `O(10^2)` rounding error margin. 

Now in the `test_transformerdecoder`, `rtol` is unchanged `1e-7`, but `atol` scales with the input, if the inputs are from 10-20, then `atol` is dialed down from `1e-5` to `1e-4`, if the inputs are in the range of 50-80, the `atol` is further reduced to `1e-3`. The relative tolerance tells the story.

@jbschlosser @zhangguanheng66 @albanD

scaomath commented on torch/nn/modules/transformer.py:
```diff
@@ -177,15 +186,22 @@ def forward(self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_ma
         """
-        output = src
+        if self.batch_first:
+            output = src.transpose(1, 0).contiguous()
```
Just my habit...do you want me to remove it?

jbschlosser replied:
Yes please- minimizing the calls to only those that are strictly necessary helps performance by avoiding unnecessary copying.

scaomath replied:
Now every `.contiguous()` is removed.

scaomath commented on torch/nn/modules/transformer.py:
```diff
@@ -318,18 +351,26 @@ class TransformerDecoderLayer(Module):
         >>> tgt = torch.rand(20, 32, 512)
         >>> out = decoder_layer(tgt, memory)
+
+    Compare with the output when `batch_first` is `True`:
```
Cool. Changed the phrasing.

jbschlosser reviewed (commented):
Updated tests are looking good! One suggested revision below for handling `batch_first` in layers only rather than at the layer container level.

Also, some tedious stuff (sorry) - it seems there are quite a few formatting issues being flagged by the linter- would you mind fixing those? Running `python -m flake8` in the root pytorch directory will identify the issues. Could you also rebase onto master to avoid merge conflicts?

scaomath commented on torch/nn/modules/transformer.py:
```diff
@@ -46,28 +48,30 @@ class Transformer(Module):
         if custom_encoder is not None:
             self.encoder = custom_encoder
         else:
             encoder_layer = TransformerEncoderLayer(d_model, nhead, dim_feedforward, dropout, activation, layer_norm_eps)
```
Sounds good. I will make the change according to your description later tonight.

Which branch should I the rebase to? The current `master`? I am seeing like a few hundreds in GitKraken...

jbschlosser replied:
> Which branch should I the rebase to? The current master? I am seeing like a few hundreds in GitKraken...

Yep, current master is the one! Thanks!

scaomath replied:
Just rebased with `master`. Now `Transformer` takes `b_f` and passes to encoder/decoder layers instead of the wrapper. Accordingly the tests in `test_nn.py` have been changed. @jbschlosser

jbschlosser reviewed (approved):
LGTM!

jbschlosser commented:
@scaomath Looks like the quantization test failures are real. Fix is to pass `batch_first` here:
https://github.com/pytorch/pytorch/blob/08561cad1014d6fec184c32529e0ec3b0a9916fd/torch/nn/quantizable/modules/activation.py#L155-L158

I missed this before, but can you also modify MHA so `batch_first` is at the end for BC reasons?

scaomath commented:
> @scaomath Looks like the quantization test failures are real. Fix is to pass `batch_first` here:
> https://github.com/pytorch/pytorch/blob/08561cad1014d6fec184c32529e0ec3b0a9916fd/torch/nn/quantizable/modules/activation.py#L155-L158
> 
> I missed this before, but can you also modify MHA so `batch_first` is at the end for BC reasons?

Makes sense, I have moved that args to the last place. The `_forward_impl()` in the MHA module in quantizable is changed accordingly.  Interestingly, the MHA in quantizable module is doing mm using batch_first by default... 😅

https://github.com/pytorch/pytorch/blob/08561cad1014d6fec184c32529e0ec3b0a9916fd/torch/nn/quantizable/modules/activation.py#L364-L368

jbschlosser commented on test/test_nn.py:
```diff
@@ -6682,69 +6685,69 @@ def perm_fn(x):
         self.assertEqual(tuple(result.shape), tuple(ref_output.shape))
```
Hm I think something got messed up in this file- shouldn't the logic be indented underneath the `batch_first` for loop?

scaomath replied:
Hmm...earlier this morning there was a conflict when merging, so I accepted all incoming changes in VSCode. I did not pay much attention and just saw `torch.Tensor` -> `torch.tensor`. Then I found there are things missing. You want to me to revert back an earlier version of `test_nn.py` that passed the test?

jbschlosser replied:
Sure, reverting sounds good! As you mentioned, something landed while this was in review to change all references of `torch.Tensor` -> `torch.tensor` in `test_nn.py`. Alongside that, constant tensors were changed to floats (e.g. `10` -> `10.`). This should be straightforward albeit tedious to fix in the reverted `test_nn.py`.

scaomath replied:
Now it should be done. Last night I was trying to do a revert but GitKraken says the action will be destructive when I tried to push, so this morning I did a manual copy-pasta. I verified the newly modified test here: 
https://github.com/scaomath/torch-playground/blob/1537ec97ade78de4f64b528cfa349acf024510e7/test_transformer.py#L1-L17
Runs no problem locally.

VitalyFedyunin commented:
Looks like this PR contains unrelated ideep and mkldnn update, it is intended? 
Anyway reverting it because of the failed bazel build https://app.circleci.com/pipelines/github/pytorch/pytorch/300968/workflows/f0f245fc-44d9-4e7c-8816-e3647c707803/jobs/12372716

jbschlosser commented:
> Looks like this PR contains unrelated ideep and mkldnn update, it is intended?
> Anyway reverting it because of the failed bazel build https://app.circleci.com/pipelines/github/pytorch/pytorch/300968/workflows/f0f245fc-44d9-4e7c-8816-e3647c707803/jobs/12372716

Ah good catch- thanks for taking care of that!

jbschlosser commented:
@scaomath when you get a chance, can you revert the ideep / mkldnn submodule change? Once that's done, we should be good to land.

scaomath committed:
update batch_first for transformers, encoder, decoder layers, MHA
update test_nn to add batch_first block, adjust test_nn atol for inputs with different magnitudes

scaomath commented:
> @scaomath when you get a chance, can you revert the ideep / mkldnn submodule change? Once that's done, we should be good to land.

@jbschlosser @VitalyFedyunin My bad...I blame vscode being too automatic, earlier it complains that it cannot load all submodules and prompted a fix, my guess is that was the culprit. I reverted the changes in submodules and rebased the commit. Now the files changed only contain `activation.py` in quantizable and the regular `nn`, `transformer.py`, and `test_nn.py`.

jbschlosser commented:
> @jbschlosser @VitalyFedyunin My bad...I blame vscode being too automatic, earlier it complains that it cannot load all submodules and prompted a fix, my guess is that was the culprit. I reverted the changes in submodules and rebased the commit. Now the files changed only contain `activation.py` in quantizable and the regular `nn`, `transformer.py`, and `test_nn.py`.

No worries- looks good now!
